File: Linkedin/LC352DataStreamasDisjointIntervals.py
# Definition for an interval.
# class Interval(object):
#     def __init__(self, s=0, e=0):
#         self.start = s
#         self.end = e

class SummaryRanges(object):

    def __init__(self):
        self.list = []

    def addNum(self, val):
        left = 0
        right = len(self.list)-1
        while left <= right:
            mid = (left+right)/2
            interval = self.list[mid]
            if interval.start <= val <= interval.end:
                return
            elif interval.start > val:
                right = mid-1
            else:
                left = mid+1
        pos = min(left, right)+1
        self.list.insert(pos, Interval(val, val))
        if pos+1 < len(self.list) and val == self.list[pos+1].start-1:
            self.list[pos].end = self.list[pos+1].end
            self.list[pos+1:pos+2] = []
        if pos-1 >= 0 and val == self.list[pos-1].end+1:
            self.list[pos-1].end = self.list[pos].end
            self.list[pos:pos+1] = []


    def getIntervals(self):
        return self.list

# Your SummaryRanges object will be instantiated and called as such:
# obj = SummaryRanges()
# obj.addNum(val)
# param_2 = obj.getIntervals()

File: apple/LC352DataStreamasDisjointIntervals.java
/**
 * Definition for an interval.
 * public class Interval {
 *     int start;
 *     int end;
 *     Interval() { start = 0; end = 0; }
 *     Interval(int s, int e) { start = s; end = e; }
 * }
 */
class SummaryRanges {
    List<Interval> array;
    public SummaryRanges() {
        array = new ArrayList<>();
    }
    public void addNum(int val) {
        int left = 0;
        int right = array.size()-1;
        while (left <= right) {
            int mid = (right-left)/2 + left;
            Interval interval = array.get(mid);
            if (interval.start <= val && val <= interval.end) {
                return;
            } else if (interval.start > val) {
                right = mid-1;
            } else {
                left = mid+1;
            }
        }
        int position = Math.min(left, right)+1;
        array.add(position, new Interval(val, val));
        if (position+1 < array.size() && val == array.get(position+1).start-1) {
            array.get(position).end = array.get(position+1).end;
            array.remove(position+1);
        } 
        if (position-1 >= 0 && val == array.get(position-1).end+1) {
            array.get(position).start = array.get(position-1).start;
            array.remove(position-1);
        }
    }
    public List<Interval> getIntervals() {
        return array;
    }
}

/**
 * Your SummaryRanges object will be instantiated and called as such:
 * SummaryRanges obj = new SummaryRanges();
 * obj.addNum(val);
 * List<Interval> param_2 = obj.getIntervals();
 */